src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs

```diff
@@ -20331,616 +20337,286 @@ static void Main()
+            var error2Symbol = comp.GetTypeByMetadataName("Error2");
+            Assert.Equal(new[] { "I<(System.Int32 notA, System.Int32 notB)>" },
+                error2Symbol.AllInterfaces.Select(i => i.ToTestDisplayString()));
 
-            CompileAndVerify(comp, expectedOutput: "hello");
+            var error3Symbol = comp.GetTypeByMetadataName("Error3");
+            Assert.Equal(new[] { "I<(System.Int32 stillNotA, System.Int32 stillNotB)>" },
+                error3Symbol.AllInterfaces.Select(i => i.ToTestDisplayString()));
+
+            var error4Symbol = comp.GetTypeByMetadataName("Error4");
+            Assert.Equal(new[] { "I<(System.Int32, System.Int32)>" },
+                error4Symbol.AllInterfaces.Select(i => i.ToTestDisplayString()));
         }
 
         [Fact]
         [WorkItem(14841, "https://github.com/dotnet/roslyn/issues/14841")]
-        public void GenericExplicitIEnumerableImplementationUsedWithDifferentTupleNames()
+        public void ChangeTupleNamesFromBaseViaGeneric()
         {
             var source = @"
-using System;
-using System.Collections;
-using System.Collections.Generic;
-
-class Base : IEnumerable<(int a, int b)>
-{
-    IEnumerator<(int a, int b)> IEnumerable<(int a, int b)>.GetEnumerator() { throw new Exception(); }
-    IEnumerator IEnumerable.GetEnumerator() { throw new Exception(); }
-}
-
-class Derived<T> : Base, IEnumerable<T>
+interface I<T> { }
+class Base : I<(int a, int b)> { }
+class Generic<U> : Base, I<U> { }
+class Error1 : Generic<(int notA, int notB)> { }
+class Error2<E2> where E2 : Generic<(int notA, int notB)> { }
+class C
 {
-    public T state;
-    IEnumerator<T> IEnumerable<T>.GetEnumerator() { return new DerivedEnumerator() { state = state }; }
-    IEnumerator IEnumerable.GetEnumerator() { throw new Exception(); }
-
-    public class DerivedEnumerator : IEnumerator<T>
+    static Generic<(int notA, int notB)> MError3()
     {
-        public T state;
-        bool done = false;
-        public T Current { get { return state; } }
-        public bool MoveNext() { if (done) { return false; } else { done = true; return true; } }
-        public void Reset() { }
-        public void Dispose() { }
-        object IEnumerator.Current { get { throw new Exception(); } }
+        return null;
     }
-}
-
-class C
-{
-    static void Main()
+    static void M()
     {
-        var collection = new Derived<(int notA, int notB)>() { state = (42, 43) };
-        foreach (var x in collection)
-        {
-            System.Console.WriteLine(x.notA);
-        }
+        I<Generic<(int notA, int notB)>> error4 = null;
+        var error5 = new Generic<(int notA, int notB)>();
+        M6((notA: 1, notB: 2)); // caught during final validation
+        M7(null, (notA: 1, notB: 2)); // candidate discarded during overload resolution (constraint checks on each parameter after substitution)
+        System.Console.WriteLine($""{error4}"");
+    }
+    static Generic<T> M6<T>(T t)
```

Comment:
Yes, when the problem appears not at the top after the substitution happens.

---
In reply to: [92057413](https://github.com/dotnet/roslyn/pull/15809#discussion_r92057413) [](ancestors = 92057413,92056785,92051906)